File: YacBoard/FEN.go
package YacBoard

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

const FenMaxLength = (Width+1)*Height + 20

func (board *YacBoard) GetFEN() string {
	result := make([]byte, 0, 64)

	for y := 0; y < Height; y++ {
		result = append(result, '/')
		for x := 0; x < Width; x++ {
			c := PieceToChar(board.GetFieldXY(x, y))
			if c == '.' {
				if unicode.IsDigit(rune(result[len(result)-1])) {
					result[len(result)-1]++
				} else {
					result = append(result, '1')
				}
			} else {
				result = append(result, c)
			}
		}
	}

	if board.WhiteMove {
		result = append(result, ' ', 'w', ' ')
	} else {
		result = append(result, ' ', 'b', ' ')
	}

	if board.WhiteCanCastleKingside {
		result = append(result, 'K')
	}
	if board.WhiteCanCastleQueenside {
		result = append(result, 'Q')
	}
	if board.BlackCanCastleKingside {
		result = append(result, 'k')
	}
	if board.BlackCanCastleQueenside {
		result = append(result, 'q')
	}
	if !board.WhiteCanCastleKingside && !board.WhiteCanCastleQueenside && !board.BlackCanCastleKingside && !board.BlackCanCastleQueenside {
		result = append(result, '-')
	}

	result = append(result, fmt.Sprintf(" %s %d %d", PosToChars(board.EnPassantPos), board.HalfmoveClock, board.MoveNumber)...)

	return string(result[1:])
}

func (board *YacBoard) SetFEN(fen string) error {
	board.Clear()

	if len(fen) > FenMaxLength {
		return errors.New("invalid FEN: too long")
	}

	splits := strings.Split(strings.TrimSpace(fen), " ")
	if len(splits) != 6 {
		if len(splits) < 6 {
			return errors.New("invalid FEN: too short")
		} else {
			return errors.New("invalid FEN: too long")
		}
	}
	lines := strings.Split(splits[0], "/")
	if len(lines) != Height {
		return errors.New(fmt.Sprintf("invalid FEN: ranks: %d, expected: %d", len(lines), Height))
	}

	// --- 1 / 6 - read pieces ---

	for y := 0; y < len(lines); y++ {
		x := 0
		for _, c := range lines[y] {
			piece := PieceFromChar(c)
			if piece == Blocked {
				return errors.New(fmt.Sprintf("invalid FEN: unknown char: %v", strconv.QuoteRune(c)))
			}
			if piece == PieceNone {
				x += int(c - '0')
				continue
			}
			board.SetField(x+y*Width, piece)
			x++
		}

		if x != Width {
			return errors.New(fmt.Sprintf("invalid FEN: at rank %d: files: %d, expected: %d", Height-y, x, Width))
		}
	}

	// --- 2 / 6 - side ---
	switch splits[1] {
	case "w":
		board.WhiteMove = true
	case "b":
		board.WhiteMove = false
	default:
		return errors.New(fmt.Sprintf("invalid FEN: unknown move color: \"%v\"", splits[1]))
	}

	// --- 3 / 6 - castling opportunities ---
	switch splits[2] {
	case "-":
	case "q":
		board.BlackCanCastleQueenside = true
	case "k":
		board.BlackCanCastleKingside = true
	case "kq":
		board.BlackCanCastleKingside = true
		board.BlackCanCastleQueenside = true
	case "Q":
		board.WhiteCanCastleQueenside = true
	case "Qq":
		board.WhiteCanCastleQueenside = true
		board.BlackCanCastleQueenside = true
	case "Qk":
		board.WhiteCanCastleQueenside = true
		board.BlackCanCastleKingside = true
	case "Qkq":
		board.WhiteCanCastleQueenside = true
		board.BlackCanCastleKingside = true
		board.BlackCanCastleQueenside = true
	case "K":
		board.WhiteCanCastleKingside = true
	case "Kq":
		board.WhiteCanCastleKingside = true
		board.BlackCanCastleQueenside = true
	case "Kk":
		board.WhiteCanCastleKingside = true
		board.BlackCanCastleKingside = true
	case "Kkq":
		board.WhiteCanCastleKingside = true
		board.BlackCanCastleKingside = true
		board.BlackCanCastleQueenside = true
	case "KQ":
		board.WhiteCanCastleKingside = true
		board.WhiteCanCastleQueenside = true
	case "KQq":
		board.WhiteCanCastleKingside = true
		board.WhiteCanCastleQueenside = true
		board.BlackCanCastleQueenside = true
	case "KQk":
		board.WhiteCanCastleKingside = true
		board.WhiteCanCastleQueenside = true
		board.BlackCanCastleKingside = true
	case "KQkq":
		board.WhiteCanCastleKingside = true
		board.WhiteCanCastleQueenside = true
		board.BlackCanCastleKingside = true
		board.BlackCanCastleQueenside = true
	default:
		return errors.New(fmt.Sprintf("invalid FEN: unknown castling: \"%v\"", splits[2]))
	}

	// --- 4 / 6 - "en passant" ---
	board.EnPassantPos = PosFrom2Chars(splits[3])
	if board.EnPassantPos == -1 && splits[3] != "-" {
		return errors.New(fmt.Sprintf("invalid FEN: unknown en passant value: \"%v\"", splits[3]))
	}

	// --- 5 / 6 - read halfmove clock ---
	tmp, err := strconv.Atoi(splits[4])
	if err != nil || tmp < 0 || tmp > 9999 {
		return errors.New(fmt.Sprintf("invalid FEN: halfmove-counter: \"%v\"", splits[4]))
	}
	board.HalfmoveClock = tmp

	// --- 6 / 6 - read move number ---
	tmp, err = strconv.Atoi(splits[5])
	if err != nil || tmp < 1 || tmp > 9999 {
		return errors.New(fmt.Sprintf("invalid FEN: movenumber: \"%v\"", splits[4]))
	}
	board.MoveNumber = tmp

	return nil
}

func (board *YacBoard) GetFastFEN(buf []byte, ofs int) int {
	p := 0
	gap := 0
	for _, field := range board.Fields {
		if field == PieceNone {
			gap++
			continue
		}
		if gap > 0 {
			buf[ofs+p] = byte(uint(gap))
			p++
			gap = 0
		}
		buf[ofs+p] = byte(field)
		p++
	}
	if gap > 0 {
		buf[ofs+p] = byte(uint(gap))
		p++
	}

	binfo := uint32(board.GetBoardInfo())
	buf[ofs+p] = byte(binfo)
	buf[ofs+p+1] = byte(binfo >> 8)
	buf[ofs+p+2] = byte(binfo >> 16)
	buf[ofs+p+3] = byte(binfo >> 24)
	buf[ofs+p+4] = byte(board.MoveNumber)
	buf[ofs+p+5] = byte(board.MoveNumber >> 8)
	p += 6

	return p
}

func (board *YacBoard) SetFastFEN(buf []byte, ofs int) int {
	p := 0
	var b byte
	for i := 0; i < len(board.Fields); i++ {
		b = buf[ofs+p]
		p++
		if b < 64 { // gap?
			board.Fields[i] = PieceNone
			for b > 1 {
				i++
				board.Fields[i] = PieceNone
				b--
			}
			continue
		}
		board.Fields[i] = Piece(b)
		if Piece(b)&King != PieceNone {
			if Piece(b) == WhiteKing {
				board.WhiteKingPos = i
			} else {
				board.BlackKingPos = i
			}
		}
	}
	var binfo = uint32(buf[ofs+p]) | uint32(buf[ofs+p+1])<<8 | uint32(buf[ofs+p+2])<<16 | uint32(buf[ofs+p+3])<<24
	board.SetBoardInfo(BoardInfo(binfo))
	board.MoveNumber = int(buf[ofs+p+4]) | int(buf[ofs+p+5])<<8
	p += 6
	return 0
}

File: dart/yacboard/lib/yacboard/fen.dart
import 'package:yacboard/yacboard/board_info.dart';

import 'board_size.dart';
import 'piece.dart';
import 'pos.dart';
import 'yac_board.dart';

extension Fen on YacBoard {
  bool setFen(String fen) {
    clear();

    if (fen.length > BoardSize.fenMaxBytes) return false; // fen is too long

    var splits = fen.split(" ");
    if (splits.length != 6) return false; // element-count != 6

    var lines = splits[0].split("/");
    if (lines.length != BoardSize.height) return false; // invalid board height

    // --- 1 / 6 - read pieces ---
    for (int y = 0; y < lines.length; y++) {
      var line = lines[y];
      int x = 0;
      for (int i = 0; i < line.length; i++) {
        int p = Piece.fromChar(line[i]);
        if (p == Piece.blocked) return false; // unknown char
        if (p == Piece.none) {
          x += line.codeUnitAt(i) - 48; // 48 = "0"
          continue;
        }
        if (x < BoardSize.width) {
          setField(Pos.fromXY(x, y), p);
        }
        x++;
      }

      if (x != BoardSize.width) return false; // invalid width
    }

    // --- 2 / 6 - side ---
    switch (splits[1]) {
      case "w":
        whiteMove = true;
        break;
      case "b":
        whiteMove = false;
        break;
      default:
        return false; // unknown move color
    }

    // --- 3 / 6 - castling opportunities ---
    switch (splits[2]) {
      case "-":
        break;
      case "q":
        blackCanCastleQueenside = true;
        break;
      case "k":
        blackCanCastleKingside = true;
        break;
      case "kq":
        blackCanCastleKingside = true;
        blackCanCastleQueenside = true;
        break;
      case "Q":
        whiteCanCastleQueenside = true;
        break;
      case "Qq":
        whiteCanCastleQueenside = true;
        blackCanCastleQueenside = true;
        break;
      case "Qk":
        whiteCanCastleQueenside = true;
        blackCanCastleKingside = true;
        break;
      case "Qkq":
        whiteCanCastleQueenside = true;
        blackCanCastleKingside = true;
        blackCanCastleQueenside = true;
        break;
      case "K":
        whiteCanCastleKingside = true;
        break;
      case "Kq":
        whiteCanCastleKingside = true;
        blackCanCastleQueenside = true;
        break;
      case "Kk":
        whiteCanCastleKingside = true;
        blackCanCastleKingside = true;
        break;
      case "Kkq":
        whiteCanCastleKingside = true;
        blackCanCastleKingside = true;
        blackCanCastleQueenside = true;
        break;
      case "KQ":
        whiteCanCastleKingside = true;
        whiteCanCastleQueenside = true;
        break;
      case "KQq":
        whiteCanCastleKingside = true;
        whiteCanCastleQueenside = true;
        blackCanCastleQueenside = true;
        break;
      case "KQk":
        whiteCanCastleKingside = true;
        whiteCanCastleQueenside = true;
        blackCanCastleKingside = true;
        break;
      case "KQkq":
        whiteCanCastleKingside = true;
        whiteCanCastleQueenside = true;
        blackCanCastleKingside = true;
        blackCanCastleQueenside = true;
        break;
      default:
        return false; // unknown castling
    }

    // --- 4 / 6 - "en passant" ---
    enPassantPos = Pos.fromChars(splits[3]);
    if (enPassantPos > 0) {
      if (whiteMove) {
        if (enPassantPos < Pos.fromChars("a6") || enPassantPos > Pos.fromChars("h6")) {
          enPassantPos = BoardInfo.EnPassantNone;
        }
        if (enPassantPos > 0 && fields[enPassantPos + BoardSize.width] != Piece.blackPawn) {
          enPassantPos = BoardInfo.EnPassantNone;
        }
      } else {
        if (enPassantPos < Pos.fromChars("a3") || enPassantPos > Pos.fromChars("h3")) {
          enPassantPos = BoardInfo.EnPassantNone;
        }
        if (enPassantPos > 0 && fields[enPassantPos - BoardSize.width] != Piece.whitePawn) {
          enPassantPos = BoardInfo.EnPassantNone;
        }
      }
    } else {
      enPassantPos = BoardInfo.EnPassantNone;
    }

    if (enPassantPos == BoardInfo.EnPassantNone && splits[3] != "-") return false; // invalid en passant value

    // --- 5 / 6 - read halfmove clock ---
    halfmoveClock = int.tryParse(splits[4]) ?? -1;
    if (halfmoveClock < 0) return false; // invalid halfmove-counter

    // --- 6 / 6 - read move number ---
    moveNumber = int.tryParse(splits[5]) ?? -1;
    if (moveNumber < 1) return false; // invalid movenumber

    return true;
  }

  String getFen() {
    List<int> result = [];

    for (int y = 0; y < BoardSize.height; y++) {
      result.add(47); // '/'
      for (int x = 0; x < BoardSize.width; x++) {
        var c = Piece.toChar(getField(Pos.fromXY(x, y)));
        if (c == ".") {
          // > '0' && < '9'
          if (result[result.length - 1] > 48 && result[result.length - 1] < 57) {
            result[result.length - 1]++;
          } else {
            result.add(49); // '1'
          }
        } else {
          result.add(c.codeUnitAt(0));
        }
      }
    }

    if (whiteMove) {
      result.addAll(" w ".codeUnits);
    } else {
      result.addAll(" b ".codeUnits);
    }

    if (whiteCanCastleKingside) result.add(75); // 'K'
    if (whiteCanCastleQueenside) result.add(81); // 'Q'
    if (blackCanCastleKingside) result.add(107); // 'k'
    if (blackCanCastleQueenside) result.add(113); // 'q'
    if (!whiteCanCastleKingside && !whiteCanCastleQueenside && !blackCanCastleKingside && !blackCanCastleQueenside) result.add(45); // '-'

    result.add(32); // ' '
    result.addAll(Pos.asString(enPassantPos).codeUnits);
    result.add(32); // ' '
    result.addAll(halfmoveClock.toString().codeUnits);
    result.add(32); // ' '
    result.addAll(moveNumber.toString().codeUnits);

    result.removeAt(0);
    return String.fromCharCodes(result);
  }
}
